sycl/doc/extensions/experimental/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -251,60 +242,82 @@ The following group algorithms support `ballot_group`, `fixed_size_group`,
 * `joint_inclusive_scan` and `inclusive_scan_over_group`
 
 
-=== Ballot-Groups
+=== Fragment
 
-A ballot-group is a non-contiguous subset of a group, representing a collection
-of all work-items in the group that share the same value of some predicate.
-Ballot-groups are always created in a range of two: the first ballot-group
-contains all work-items where the predicate is true, and the second
-ballot-group contains all work-items where the predicate is false.
+A `fragment` represents a non-contiguous subset of the work-items in some
+parent group.
 
 
 ==== Creation
 
-New ballot-groups are created by partitioning an existing group, using the
-`get_ballot_group` free-function.
+A new `fragment` can be created via the following:
 
-NOTE: Creating a ballot-group requires a barrier across all work-items in the
-parent group, since work-items must exchange predicate values in order to
-determine group membership.
+- Partitioning an existing group with the `binary_partition()` function.
+- By calling the `get_opportunistic_group()` function.
 
 [source, c++]
 ----
 namespace ext::oneapi::experimental {
 
-template <typename Group>
-ballot_group get_ballot_group(Group group, bool predicate);
+template <typename ParentGroup>
+fragment<ParentGroup> binary_partition(ParentGroup parent, bool predicate);
 
 } // namespace ext::oneapi::experimental
 ----
-_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>> &&
-std::is_same_v<Group, sycl::sub_group>` is true.
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<ParentGroup>>`
+and `ParentGroup` is a `sycl::sub_group` or `sycl::fragment`.
 
-_Preconditions_: All work-items in `group` must encounter this function in
+_Preconditions_: All work-items in `parent` must encounter this function in
 converged control flow.
 
-_Effects_: Synchronizes all work-items in `group`.
+_Effects:_ Blocks until all work-items in `parent` have reached this
+synchronization point.
 
-_Returns_: A `ballot_group` consisting of the work-items in `group` for which
-`predicate` has the same value as the calling work-item.
+_Synchronization:_ The call in each work-item happens before any work-item
+blocking on the same synchronization point is unblocked.
+Synchronization operations used by an implementation must respect the memory
+scope reported by `ParentGroup::fence_scope`.
 
+_Returns_: A `fragment` consisting of the work-items in `parent` for which
+`predicate` has the same value as the calling work-item. This `fragment`
+must have a group range of 2, and a group index of either 0 (if `predicate` is
+false) or 1 (if `predicate` is true).
 
-==== `ballot_group` Class
+
+[source, c++]
+----
+namespace ext::oneapi::experimental::this_kernel {
+
+fragment<sycl::sub_group> get_opportunistic_group();
+
+} // namespace ext::oneapi::experimental::this_kernel
+----
+
+_Returns_: A `fragment` consisting of all work-items which call this function
```

Comment:
Should we clarify that all work-items from a sub-group the caller work-item belongs to?
Or do we consider that obvious by the fact that return type is specialized on `sub_group`?

Related to a comment above: do we ever expect this function to be extended beyond sub-groups and if so, should it be templated on a parent group (which can for now be limited to `sub_group` only)?

Reply:
Done in c41c4bed593870f41dfc675d16e6e54555082d8c.

I don't think we expect this function to be extended beyond sub-groups, which is why I made this simplification.  Some devices run the work-items within a sub-group "together" within a single hardware thread (i.e., in SIMD or similar) and it makes sense to be able to query which work-items are (for lack of a better word) "active" within a hardware thread.  Identifying which work-items are active should always be inexpensive, because the hardware thread and/or hardware already have that information somewhere.

Building an opportunistic group across sub-groups would almost always require communication between different hardware threads, so it would always be expensive.  It's also _much_ harder to define what it means for two hardware threads to encounter this function call "together" (e.g., does it imply that threads wait a certain amount of time to see if other threads show up?) and I don't think we want to open that can of worms.